x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptFieldMapper.java

```diff
@@ -6,42 +6,60 @@
 
 package org.elasticsearch.xpack.runtimefields.mapper;
 
-import org.apache.lucene.document.FieldType;
-import org.apache.lucene.util.SetOnce;
 import org.elasticsearch.common.xcontent.XContentBuilder;
-import org.elasticsearch.common.xcontent.support.XContentMapValues;
 import org.elasticsearch.index.mapper.FieldMapper;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.Mapper;
 import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.ParametrizedFieldMapper;
 import org.elasticsearch.index.mapper.ParseContext;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.painless.PainlessScriptEngine;
 import org.elasticsearch.script.Script;
-import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.ScriptType;
 import org.elasticsearch.xpack.runtimefields.StringScriptFieldScript;
 
 import java.io.IOException;
-import java.util.Iterator;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Supplier;
 
-public final class ScriptFieldMapper extends FieldMapper {
+public final class ScriptFieldMapper extends ParametrizedFieldMapper {
 
     public static final String CONTENT_TYPE = "script";
 
-    private static final FieldType FIELD_TYPE = new FieldType();
+    private final String runtimeType;
+    private final Script script;
+    private final Supplier<QueryShardContext> queryShardContextSupplier;
 
-    ScriptFieldMapper(String simpleName, MappedFieldType mappedFieldType, MultiFields multiFields, CopyTo copyTo) {
-        super(simpleName, FIELD_TYPE, mappedFieldType, multiFields, copyTo);
+    private static ScriptFieldMapper toType(FieldMapper in) {
+        return (ScriptFieldMapper) in;
+    }
+
+    protected ScriptFieldMapper(
+        String simpleName,
+        MappedFieldType mappedFieldType,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        String runtimeType,
+        Script script,
+        Supplier<QueryShardContext> queryShardContextSupplier
+    ) {
+        super(simpleName, mappedFieldType, multiFields, copyTo);
+        this.runtimeType = runtimeType;
+        this.script = script;
+        this.queryShardContextSupplier = queryShardContextSupplier;
```

Comment:
I worry about keeping the `queryShardContextSupplier` around because it is sneaky. Could `getMergeBuilder` get the parse context or something like that?

Reply:
I agree. I personally prefer the hack that we had before (SetOnce in the type parser, set from the plugin class). Otherwise I can get the query shard context from the parser context and pass in only the script service, hopefully it is available.